### ISSUE TITLE
 `observe` stream may be cancelled for successful output streams

### ISSUE THREAD

**Issue by zainab-ali** [typelevel/fs2#2757]
This was raised following [a discussion in discord](https://discord.com/channels/632277896739946517/632310980449402880/920359666960171018).

The completion behaviour of `observe` has changed to be less deterministic.

```scala
outputStream = inputStream.observe(observerPipe)
```

`observe` constructs a stream from the `inputStream` and `observerPipe` (let's term this the observer stream) and runs this concurrently to the `outputStream`.

Previously, the observer stream would succeed if the output stream succeeded.  In recent fs2 versions, slow observer streams are cancelled when the output stream completes.

## Reproduction

The following code snippet shows an example of this:

```scala
val observer: Pipe[IO, Unit, Nothing] = _.chunkN(2)
  .evalMap(_ => IO.sleep(1.second))
  .onFinalizeCase(c => IO(println(s"Observer has finalized with case $c")))
  .drain

val main: Stream[IO, Unit] = Stream
  .eval(IO.unit)
  .observe(observer)
  .onFinalizeCase(c => IO(println(s"Main has finalized with case $c")))
}
```

This would previously print

```
Observer has finalized with case Completed
Main has finalized with case Completed
```

but now prints

```
Observer has finalized with case Canceled
Main has finalized with case Succeeded
```

You can experiment with this in Scastie using [v2.5.10](https://scastie.scala-lang.org/zainab-ali/n2IzsSJYTgmwkMjVaWt40w) and [v3.2.3](https://scastie.scala-lang.org/zainab-ali/CdXGdfRJSsKeW3SnhIK6NA) respectively.

## Cause

This was introduced in fs2 v3.0.0 in a refactoring of none-terminated queues. 

In [v2.5.10](https://github.com/typelevel/fs2/blob/v2.5.10/core/shared/src/main/scala/fs2/Stream.scala#L3746) the output stream was backed by a none-terminated queue. This was only terminated after the observer stream had completed, meaning that the observer stream would never be cancelled by the concurrent logic.

In [the current implementation](https://github.com/typelevel/fs2/blob/main/core/shared/src/main/scala/fs2/Stream.scala#L3828) we terminate the output stream once elements have been *pulled* by the observer stream, but not processed. The observer stream can be cancelled while processing the last elements.

I suspect this was an inadvertent change introduced [in the following refactor](https://github.com/typelevel/fs2/commit/ab2eaf387ec92ac149bdfaac2f54ca6d19b6b420).

## Concerns

There are a few concerns with the current behaviour:
 1. This logic change may affect anyone using `observe` upgrading from CE2 to CE3. It would be difficult for them to identify why the logic has changed. 
    We could address this by adding a mention in the release notes, or a migration guide.
 2. `observe` can't be used for necessary computations because cancellation is in the happy path of the observe stream.

It would be fairly easy to change the logic back to the original by closing the output channel after the observer pipe completes. 

How would you like to address this? I'm happy to make the code or docs changes myself.

**Comment by SystemFw** [comment]
I think I'd like to see the old behaviour back unless someone offers a compelling reason against it

**Comment by mpilquist** [comment]
Agreed that the old behavior would be better. Thanks for tracking this down!

**Comment by zainab-ali** [comment]
Thanks! I'll make a start on a PR.

**Comment by zainab-ali** [comment]
This was fixed in #2765.
